adding conditional execution in ingest pipeline docs (#9625)
opensearch-project/documentation-website: _ingest-pipelines/conditional-execution.md

---
layout: default
title: Conditional execution
nav_order: 40
---

# Conditional execution

> In ingest pipelines, you can control whether a processor runs by using the optional `if` parameter. This allows for conditional execution of processors based on the contents of the incoming document. The condition is written as a Painless script and evaluated against the document context (`ctx`).

Reviewer: ```suggestion
In ingest pipelines, you can control whether a processor runs by using the optional `if` parameter. This allows for conditional execution of processors based on the incoming document contents. The condition is written as a Painless script and evaluated against the document context (`ctx`).
```

## Basic conditional execution

> Each processor can include an `if` clause. If the condition evaluates to `true`, the processor runs, otherwise, it's skipped.

Reviewer 2: ```suggestion
Each processor can include an `if` clause. If the condition evaluates to `true`, the processor runs; otherwise, it's skipped.
```

### Example: Drop debug-level logs

> The following pipeline drops any document where the `log_level` field is `"debug"`.

Reviewer: ```suggestion
The following pipeline drops any document in which the `log_level` field is equal to `debug`:
```

```json
PUT _ingest/pipeline/drop_debug_logs
{
  "processors": [
    {
      "drop": {
        "if": "ctx.log_level == 'debug'"
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Example index request

```json
POST logs/_doc/1?pipeline=drop_debug_logs
{
  "message": "User logged in",
  "log_level": "debug"
}
```
{% include copy-curl.html %}

This document is dropped because the condition evaluates to `true`:

```json
{
  "_index": "logs",
  "_id": "1",
  "_version": -3,
  "result": "noop",
  "_shards": {
    "total": 0,
    "successful": 0,
    "failed": 0
  }
}
```

> ## Null-safe field checks with nested fields

Reviewer: ```suggestion
## Null-safe field checks when using nested fields
```

When working with nested fields, it's important to avoid null pointer exceptions. Use the null-safe operator `?.` in Painless scripts.

> ### Example: Drop documents based on nested field

Reviewer: ```suggestion
### Example: Drop documents based on a nested field
```

The following drop processor executes only if the nested `app.env` field exists and equals `debug`:

```json
PUT _ingest/pipeline/drop_debug_env
{
  "processors": [
    {
      "drop": {
        "if": "ctx.app?.env == 'debug'"
      }
    }
  ]
}
```
{% include copy-curl.html %}

> If the null-safe `?.` operator is not configured, indexing any document which doesn't contain `app.env` field will trigger the following null pointer exception:

Reviewer 2: ```suggestion
If the null-safe `?.` operator is not configured, indexing any document that doesn't contain `app.env` field will trigger the following null pointer exception:
```

> 

Reviewer 2: Line 82: Either "a" or "the" should precede `app.env`.

```json
{
  "error": "IngestProcessorException[ScriptException[runtime error]; nested: NullPointerException[Cannot invoke \"Object.getClass()\" because \"callArgs[0]\" is null];]",
  "status": 400
}
```

## Handling flattened fields

> If your document has a flattened field, for example: `"app.env": "debug"`, use the [`dot_expander`]({{site.url}}{{site.baseurl}}/ingest-pipelines/processors/dot-expander/) processor to convert it into a nested structure:

Reviewer 2: ```suggestion
If your document has a flattened field, for example, `"app.env": "debug"`, use the [`dot_expander`]({{site.url}}{{site.baseurl}}/ingest-pipelines/processors/dot-expander/) processor to convert it into a nested structure:
```

```json
PUT _ingest/pipeline/drop_debug_env
{
  "processors": [
    {
      "dot_expander": {
        "field": "app.env"
      }
    },
    {
      "drop": {
        "if": "ctx.app?.env == 'debug'"
      }
    }
  ]
}
```
{% include copy-curl.html %}

## Safe method calls in conditions

Avoid calling methods on potential null values. Use constants or null checks instead:

```json
{
  "drop": {
    "if": "ctx.app?.env != null && ctx.app.env.contains('debug')"
  }
}
```

## Full example: Multi-step conditional pipeline

The following ingest pipeline uses three processors:

1. `set`: If no value is provided in the `user` field, sets the `user` field to `guest`.

> 2. `set`: If the `status_code` is provided and is higher than `400`, sets the `error` field to `true` .

Reviewer 2: ```suggestion
2. `set`: If the `status_code` is provided and is higher than `400`, sets the `error` field to `true`.
```
